SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -459,30 +480,55 @@ void USpatialNetDriver::CreateServerSpatialOSNetConnection()
 }
 
-void USpatialNetDriver::QueryGSMToLoadMap()
+bool USpatialNetDriver::CanSendPlayerSpawnRequests()
 {
-	check(bConnectAsClient);
+	return GlobalStateManager->GetAcceptingPlayers() && SessionId == GlobalStateManager->GetSessionId();
+}
 
-	// Register our interest in spawning.
-	bWaitingForAcceptingPlayersToSpawn = true;
+void USpatialNetDriver::OnQueryGSMSuccess()
+{
+	// If the deployment is now accepting players and we are waiting to spawn. Spawn.
+	if (bWaitingToSpawn && CanSendPlayerSpawnRequests())
+	{
+		UWorld* CurrentWorld = GetWorld();
+		const FString& DeploymentMapURL = GlobalStateManager->GetDeploymentMapURL();
+		if (CurrentWorld == nullptr || CurrentWorld->RemovePIEPrefix(DeploymentMapURL) != CurrentWorld->RemovePIEPrefix(CurrentWorld->URL.Map))
+		{
+			// Load the correct map based on the GSM URL
+			UE_LOG(LogSpatial, Log, TEXT("Welcomed by SpatialOS (Level: %s)"), *DeploymentMapURL);
 
-	// Begin querying the state of the GSM so we know the state of AcceptingPlayers.
-	GlobalStateManager->QueryGSM(true /*bRetryUntilAcceptingPlayers*/);
+			// Extract map name and options
+			FWorldContext& WorldContext = GEngine->GetWorldContextFromPendingNetGameNetDriverChecked(this);
+			FURL LastURL = WorldContext.PendingNetGame->URL;
+
+			FURL RedirectURL = FURL(&LastURL, *DeploymentMapURL, (ETravelType)WorldContext.TravelType);
+			RedirectURL.Host = LastURL.Host;
+			RedirectURL.Port = LastURL.Port;
+			RedirectURL.AddOption(*SpatialConstants::ClientsStayConnectedURLOption);
+
+			WorldContext.PendingNetGame->bSuccessfullyConnected = true;
+			WorldContext.PendingNetGame->bSentJoinRequest = false;
+			WorldContext.PendingNetGame->URL = RedirectURL;
+		}
+		else
+		{
+			MakePlayerSpawnRequest();
+		}
+	}
 }
 
-void USpatialNetDriver::HandleOngoingServerTravel()
+void USpatialNetDriver::QueryGSMToLoadMap()
 {
-	check(!bConnectAsClient);
+	check(bConnectAsClient);
 
-	// Here if we are a server and this is server travel (there is a snapshot to load) we want to load the snapshot.
-	if (!ServerConnection && !SnapshotToLoad.IsEmpty() && Cast<USpatialGameInstance>(GetWorld()->GetGameInstance())->bResponsibleForSnapshotLoading)
-	{
-		UE_LOG(LogSpatialOSNetDriver, Log, TEXT("Worker authoriative over the GSM is loading snapshot: %s"), *SnapshotToLoad);
-		SnapshotManager->LoadSnapshot(SnapshotToLoad);
+	// Register our interest in spawning.
+	bWaitingToSpawn = true;
 
-		// Once we've finished loading the snapshot we must update our bResponsibleForSnapshotLoading in-case we do not gain authority over the new GSM.
-		Cast<USpatialGameInstance>(GetWorld()->GetGameInstance())->bResponsibleForSnapshotLoading = false;
-	}
+	UGlobalStateManager::QuerySuccessDelegate QuerySuccessDelegate;
+	QuerySuccessDelegate.BindUObject(this, &USpatialNetDriver::OnQueryGSMSuccess);
+
+	// Begin querying the state of the GSM so we know the state of AcceptingPlayers.
+	GlobalStateManager->QueryGSM(true, SessionId, QuerySuccessDelegate);
```

Comment:
```suggestion
	GlobalStateManager->QueryGSM(true /*bRetryUntilAcceptingPlayers*/, SessionId, QuerySuccessDelegate);
```
Nit: For unnamed variables we usually add the name in the function call.